### PR TITLE
Fix cancelling of started-but-lang-doesn't-know workflows

### PR DESCRIPTION
**Pull request by Sushisource**
## What was changed
* Created a facility for core to generate new commands internally
* Used that to allow child wf machine to generate a cancel external wf command when cancel is called on it, and the child WF has already started
* Various bits of cleanup as I went

I am reasonably sure this is history compat, but unfortunately that comes at the cost of lang continuing to issue a cancel external command if it does know the child is already started. Otherwise, it's possible command ordering could be slightly different. Core will generate the same cancel external command in that situation now, but that could show up in the command list in a slightly different place than it does when lang does it directly. We would need to test in TS.

## Why?
Fix the associated bug. Also allows us to remove the timer backoff hack, and that could now be done totally internally in core (however, not easy to do and maintain history compat...).

## Checklist

1. Closes https://github.com/temporalio/sdk-core/issues/331
2. How was this tested:
Added UTs, existing tests

### CODE REVIEW

File: core/src/core_tests/child_workflows.rs [hunk 1]
```diff
@@ -97,3 +103,68 @@ async fn cancel_child_workflow() {
...
+    // Make sure that a resolve for the "request cancel external workflow" command does *not* appear
+    // since lang didn't actually issue one. The only job should be resolving the child workflow.
+    assert_matches!(
+        act.jobs.as_slice(),
+        [WorkflowActivationJob {
+            variant: Some(workflow_activation_job::Variant::ResolveChildWorkflowExecution(_)),
+        }]
+    );
+    // Request cancel external is technically fallible, but the only reasons relate to targeting
+    // a not-found workflow, which couldn't happen in this case.
```
Comment (bergundy) [inline core/src/core_tests/child_workflows.rs:169]: Should we test other cases like when the child workflow is already complete? What happens in this case for example if we're processing a WFT while the child completes or we already know that the child completed?

Comment (bergundy) [reply core/src/core_tests/child_workflows.rs:169]: I'd test some more scenarios

File: core/src/core_tests/child_workflows.rs [hunk 2]
```diff
@@ -97,3 +103,68 @@ async fn cancel_child_workflow() {
     wfm.process_all_activations().await.unwrap();
     wfm.shutdown().await.unwrap();
 }
+
+#[rstest::rstest]
+#[case::abandon(ChildWorkflowCancellationType::Abandon)]
+#[case::try_cancel(ChildWorkflowCancellationType::TryCancel)]
+#[case::wait_cancel_completed(ChildWorkflowCancellationType::WaitCancellationCompleted)]
```
Comment (bergundy) [inline core/src/core_tests/child_workflows.rs:110]: nit: could also test WaitCancellationRequested

File: core/src/core_tests/child_workflows.rs [hunk 3]
```diff
@@ -97,3 +103,68 @@ async fn cancel_child_workflow() {
...
+#[tokio::test]
+async fn cancel_child_workflow_lang_thinks_not_started_but_is(
```
Comment (bergundy) [inline core/src/core_tests/child_workflows.rs:112]: Test cancellation of not started child too? Is this interesting enough to test?

Comment (bergundy) [reply core/src/core_tests/child_workflows.rs:112]: Test cancellation of already completed child, core should ignore this (we do the same for activities AFAIR).

**Review by bergundy — COMMENTED** [review]
While I definitely agree this is better than what we had I wouldn't merge it before verifying it doesn't break TS.

**Comment by Sushisource** [comment]
I verified that this does not break compat when used from TS using this branch:
https://github.com/temporalio/sdk-typescript/tree/test-child-wf-changes

File: core/src/core_tests/child_workflows.rs (hunk 3 above)
Comment (Sushisource) [reply core/src/core_tests/child_workflows.rs:112]: Added

**Review by bergundy — APPROVED** [review]

**Comment by Sushisource** [comment]
Will merge after https://github.com/temporalio/sdk-core/pull/396

**Pull request merged**